pkg/operator/ceph/object/controller.go

```diff
@@ -525,43 +506,40 @@ func (r *ReconcileCephObjectStore) reconcileMultisiteCRs(cephObjectStore *cephv1
-		logger.Info("external rgw endpoint monitoring go routine already running!")
+// handle upgrade from v1.10; we need to remove the health checker bucket from the store, or
+// deleting the store will be stuck on dependents because an unknown bucket exists
+// TODO: remove this for Rook v1.12 release
+var removeDeprecatedHealthCheckBucket = func(ctx context.Context, opsCtx *AdminOpsContext, cos *cephv1.CephObjectStore) error {
+	healthCheckBucket := genHealthCheckerBucketName(string(cos.UID))
+	doPurge := true // purge all content of the bucket
+	bucket := admin.Bucket{Bucket: healthCheckBucket, PurgeObject: &doPurge}
+	remErr := opsCtx.AdminOpsClient.RemoveBucket(ctx, bucket)
+
+	if remErr == nil {
+		logger.Info("successfully deleted deprecated health checker bucket")
 		return nil
 	}
 
-	rgwChecker, err := newBucketChecker(r.context, objContext, r.client, namespacedName, &objectstore.Spec)
-	if err != nil {
-		return errors.Wrapf(err, "failed to start rgw health checker for CephObjectStore %q, will re-reconcile", namespacedName.String())
+	if errors.Is(remErr, admin.ErrNoSuchBucket) {
+		logger.Debug("deprecated health checker bucket already does not exist: NoSuchBucket")
+		return nil
 	}
 
-	logger.Infof("starting rgw health checker for CephObjectStore %q", namespacedName.String())
-	go rgwChecker.checkObjectStore(r.objectStoreContexts, channelKey)
-
-	// Set the monitoring flag so we don't start more than one go routine
-	r.objectStoreContexts[channelKey].started = true
-
-	return nil
-}
-
-// cancel monitoring. This is a noop if monitoring is not running.
-func (r *ReconcileCephObjectStore) stopMonitoring(objectstore *cephv1.CephObjectStore) {
-	channelKey := monitoringChannelKey(objectstore)
-
-	_, monitoringContextExists := r.objectStoreContexts[channelKey]
-	if monitoringContextExists {
-		// stop the monitoring routine
-		r.objectStoreContexts[channelKey].internalCancel()
-
-		// remove the monitoring routine from the map
-		delete(r.objectStoreContexts, channelKey)
+	if errors.Is(remErr, admin.ErrNoSuchKey) {
+		// ceph might return NoSuchKey than NoSuchBucket when the target bucket does not exist.
+		// then we can use GetBucketInfo() to judge the existence of the bucket.
+		// see: https://github.com/ceph/ceph/pull/44413
+		_, getErr := opsCtx.AdminOpsClient.GetBucketInfo(ctx, bucket)
+		if getErr != nil {
+			if errors.Is(getErr, admin.ErrNoSuchBucket) {
+				logger.Debug("deprecated health checker bucket already does not exist: NoSuchKey")
+				return nil
+			}
+			// both commands errored; something is up; be sure to return both errors
+			return errors.Wrapf(remErr, "failed to delete deprecated health checker bucket %q. failed to get bucket info. %v", healthCheckBucket, getErr)
+		}
+		// remove bucket failed, and bucket exists; something else went wrong with remove bucket
 	}
+
+	return errors.Wrapf(remErr, "failed to delete deprecated health checker bucket %q", healthCheckBucket)
```

Comment:
If we failed to delete the bucket, how about if we log it, but not fail the reconcile?